include/core/Layer.hpp

```diff
@@ -0,0 +1,147 @@
+
+        }
+
+        virtual ~Layer() {}
+
+        /***
+         * @brief Return the next timestep that will be processed by this layer in epoc time units
+        */
+        time_t next_timestep_epoch_time() { return simulation_time.next_timestep_epoch_time(); }
+
+
+        /***
+         * @brief Return the last timesteo that was processed by this layer in epoc time units
+        */
+        time_t current_timestep_epoch_time() { return simulation_time.get_current_epoc_time(); }
+
+
+        /***
+         * @brief Return the numeric id of this layer
+        */
+        int get_id() const { return this->description.id; }
+
+        /***
+         * @brief Return the name of this layer
+        */
+        const std::string& get_name() const { return this->description.name; }
+
+        /***
+         * @brief Return this time_step interval used for this layer
+        */
+        const double get_time_step() const { return this->description.time_step; }
+
+        /***
+         * @brief Return the units for the time_step value of this layer
+        */
+        const std::string& get_time_step_units() const { return this->description.time_step_units; }
+
+        /***
+         * @breif Run one simulation timestep for each model in this layer
+        */
+        virtual void update_models()
+        {
+            auto idx = simulation_time.next_timestep_index();
+            auto step = simulation_time.get_output_interval_seconds();
+            
+            //std::cout<<"Output Time Index: "<<output_time_index<<std::endl;
+            if(output_time_index%100 == 0) std::cout<<"Running timestep " << output_time_index <<std::endl;
+            std::string current_timestamp = simulation_time.get_timestamp(output_time_index);
+            for(const auto& id : processing_units) 
+            {
+                int sub_time = output_time_index;
+                //std::cout<<"Running cat "<<id<<std::endl;
+                auto r = features.catchment_at(id);
+                //TODO redesign to avoid this cast
+                auto r_c = std::dynamic_pointer_cast<realization::Catchment_Formulation>(r);
+                r_c->set_et_params(pdm_et_data);
+                double response = r_c->get_response(output_time_index, description.time_step);
+                std::string output = std::to_string(output_time_index)+","+current_timestamp+","+
+                                    r_c->get_output_line_for_timestep(output_time_index)+"\n";
+                r_c->write_output(output);
+                //TODO put this somewhere else.  For now, just trying to ensure we get m^3/s into nexus output
+                try{
+                    response *= (catchment_data->get_feature(id)->get_property("areasqkm").as_real_number() * 1000000);
+                }catch(std::invalid_argument &e)
+                {
+                    response *= (catchment_data->get_feature(id)->get_property("area_sqkm").as_real_number() * 1000000);
+                }
```

Comment:
It would be nice if the actual variation between the two arms (the key name) were isolated from the repeated logic.

Perhaps this whole thing should actually be encapsulated inside a member `catchment_data::get_area()` (cf [Law of Demeter](https://en.wikipedia.org/wiki/Law_of_Demeter#In_object-oriented_programming), and then it can be made to cache it so we don't incur repeated lookup costs

Reply:
The performance concern holds doubly so because `catchment_data` is a (should-be-`const`) member of this class, and so the area can be looked up once outside any timestep logic.